crates/wasi-common/src/sys/windows/hostcalls_impl/fs.rs

```diff
@@ -362,109 +380,141 @@ pub(crate) fn fd_readdir(
-                        fs::remove_dir(&new_path)
-                            .and_then(|()| fs::rename(old_path, new_path))
-                            .map_err(Into::into)
-                    }
-                }
-                WinError::ERROR_INVALID_NAME => {
-                    // If source contains trailing slashes, check if we are dealing with
-                    // a file instead of a dir, and if so, throw ENOTDIR.
-                    if let Some(path) = strip_trailing_slashes_and_concatenate(&resolved_old)? {
-                        if path.is_file() {
-                            return Err(Error::ENOTDIR);
+                    log::debug!("path_rename at rename error code={:?}", e);
+                    match WinError::from_u32(e as u32) {
+                        WinError::ERROR_ACCESS_DENIED => {
+                            // So most likely dealing with new_path == dir.
+                            // Eliminate case old_path == file first.
+                            if old_path.is_file() {
+                                Err(Error::EISDIR)
+                            } else {
+                                // Ok, let's try removing an empty dir at new_path if it exists
+                                // and is a nonempty dir.
+                                fs::remove_dir(&new_path)
+                                    .and_then(|()| fs::rename(old_path, new_path))
+                                    .map_err(Into::into)
+                            }
                         }
+                        WinError::ERROR_INVALID_NAME => {
+                            // If source contains trailing slashes, check if we are dealing with
+                            // a file instead of a dir, and if so, throw ENOTDIR.
+                            if let Some(path) =
+                                strip_trailing_slashes_and_concatenate(&resolved_old)?
+                            {
+                                if path.is_file() {
+                                    return Err(Error::ENOTDIR);
+                                }
+                            }
+                            Err(WinError::ERROR_INVALID_NAME.into())
+                        }
+                        e => Err(e.into()),
                     }
-                    Err(WinError::ERROR_INVALID_NAME.into())
                 }
-                e => Err(e.into()),
-            }
+                None => {
+                    log::debug!("Inconvertible OS error: {}", e);
+                    Err(Error::EIO)
+                }
+            })
         }
-        None => {
-            log::debug!("Inconvertible OS error: {}", e);
-            Err(Error::EIO)
+        (Descriptor::VirtualFile(_), _) | (_, Descriptor::VirtualFile(_)) => {
+            unimplemented!("path_rename with one or more virtual files");
         }
-    })
+        _ => {
+            unreachable!("streams do not have paths and should not be accessible via PathGet");
```

Comment:
Will this remain true, generally, as wasi-common grows an idea of streams outside stdin/stdout/stderr?

Reply:
Excellent question, and while I don't possess a clear-cut answer for this, I think it will *not* remain true. But then again, I think that introducing streams into `wasi-common` will require at least some rewriting of some syscalls. I don't have a problem leaving it as `unreachable`. Alternatively, I guess we could change it to `unimplemented` but I agree with your train of thoughts here, only if we're certain this *will* apply to streams as well. All in all, I'm voting for leaving that in as-is.